src/humanloop/otel/processor.py

```diff
@@ -0,0 +1,173 @@
+    # used for this
+
+    def on_end(self, span: ReadableSpan) -> None:
+        if is_humanloop_span(span=span):
+            _process_humanloop_span(span, self._children[span.context.span_id])
+            del self._children[span.context.span_id]
+            self.span_exporter.export([span])
+        else:
+            if span.parent is not None and _is_instrumentor_span(span):
+                self._children[span.parent.span_id].append(span)
+
+
+def _is_instrumentor_span(span: ReadableSpan) -> bool:
+    # TODO: Extend in the future as needed. Spans not coming from
+    # Instrumentors of interest should be dropped
+    return is_llm_provider_call(span=span)
+
+
+def _process_humanloop_span(span: ReadableSpan, children_spans: list[ReadableSpan]):
+    hl_file = read_from_opentelemetry_span(span, key=HL_FILE_OT_KEY)
+
+    if "prompt" in hl_file:
+        _process_prompt(prompt_span=span, children_spans=children_spans)
+        return
+    elif "tool" in hl_file:
+        _process_tool(tool_span=span, children_spans=children_spans)
+        return
+    elif "flow" in hl_file:
+        _process_flow(flow_span=span, children_spans=children_spans)
+        return
+    else:
+        logging.error("Invalid span type")
+
+
+def _process_prompt(prompt_span: ReadableSpan, children_spans: list[ReadableSpan]):
+    if len(children_spans) == 0:
+        return
+    child_span = children_spans[0]
+    assert is_llm_provider_call(child_span)
+    _enrich_prompt_span_file(prompt_span, child_span)
+    _enrich_prompt_span_log(prompt_span, child_span)
+
+
+def _process_tool(tool_span: ReadableSpan, children_spans: list[ReadableSpan]):
+    # TODO: Use children_spans in the future
+    tool_log = read_from_opentelemetry_span(tool_span, key=HL_LOG_OT_KEY)
+    tool_log["start_time"] = tool_span.start_time / 1e9
+    tool_log["end_time"] = tool_span.end_time / 1e9
+    tool_log["created_at"] = tool_span.end_time / 1e9
+
+    write_to_opentelemetry_span(
+        span=tool_span,
+        key=HL_LOG_OT_KEY,
+        value=tool_log,
+    )
+
+
+def _process_flow(flow_span: ReadableSpan, children_spans: list[ReadableSpan]):
+    # TODO: Use children_spans in the future
+    flow_log = read_from_opentelemetry_span(flow_span, key=HL_LOG_OT_KEY)
+    flow_log["start_time"] = flow_span.start_time / 1e9
+    flow_log["end_time"] = flow_span.end_time / 1e9
+    flow_log["created_at"] = flow_span.end_time / 1e9
+
+    write_to_opentelemetry_span(
+        span=flow_span,
+        key=HL_LOG_OT_KEY,
+        value=flow_log,
+    )
+
+
+def _enrich_prompt_span_file(prompt_span: ReadableSpan, llm_provider_call_span: ReadableSpan):
+    hl_file = read_from_opentelemetry_span(prompt_span, key=HL_FILE_OT_KEY)
+    gen_ai_object = read_from_opentelemetry_span(llm_provider_call_span, key="gen_ai")
+    llm_object = read_from_opentelemetry_span(llm_provider_call_span, key="llm")
+
+    prompt_kernel = hl_file.get("prompt", {})
```

Comment:
Are you sure this will work across providers? 

I am pretty sure that different providers have different param naming conventions

Reply:
Yes! The OpenLLMetry project is maintaining an [official extension](https://github.com/open-telemetry/community/blob/1c71595874e5d125ca92ec3b0e948c4325161c8a/projects/llm-semconv.md) of the OTel standard that defines how logging of LLM-backed features should happen.

All their [Instrumentors](https://github.com/open-telemetry/community/blob/1c71595874e5d125ca92ec3b0e948c4325161c8a/projects/llm-semconv.md) output Spans with an uniform format, so the keys are guaranteed.

I have also added tests with multiple providers to assert this.